src/plugins/action_server/include/plugins/action_server/action_server.h

```diff
@@ -0,0 +1,306 @@
+// Edits need to be made to the proto files
+// (see https://github.com/mavlink/MAVSDK-Proto/blob/main/protos/action_server/action_server.proto)
+
+#pragma once
+
+#include <array>
+#include <cmath>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "plugin_base.h"
+
+namespace mavsdk {
+
+class System;
+class ActionServerImpl;
+
+/**
+ * @brief Provide vehicle actions (as a server) such as arming, taking off, and landing.
+ */
+class ActionServer : public PluginBase {
+public:
+    /**
+     * @brief Constructor. Creates the plugin for a specific System.
+     *
+     * The plugin is typically created as shown below:
+     *
+     *     ```cpp
+     *     auto action_server = ActionServer(system);
+     *     ```
+     *
+     * @param system The specific system associated with this plugin.
+     */
+    explicit ActionServer(System& system); // deprecated
+
+    /**
+     * @brief Constructor. Creates the plugin for a specific System.
+     *
+     * The plugin is typically created as shown below:
+     *
+     *     ```cpp
+     *     auto action_server = ActionServer(system);
+     *     ```
+     *
+     * @param system The specific system associated with this plugin.
+     */
+    explicit ActionServer(std::shared_ptr<System> system); // new
+
+    /**
+     * @brief Destructor (internal use only).
+     */
+    ~ActionServer();
+
+    /**
+     * @brief Flight modes.
+     *
+     * For more information about flight modes, check out
+     * https://docs.px4.io/master/en/config/flight_mode.html.
+     */
+    enum class FlightMode {
+        Unknown, /**< @brief Mode not known. */
+        Ready, /**< @brief Armed and ready to take off. */
+        Takeoff, /**< @brief Taking off. */
+        Hold, /**< @brief Holding (hovering in place (or circling for fixed-wing vehicles). */
+        Mission, /**< @brief In mission. */
+        ReturnToLaunch, /**< @brief Returning to launch position (then landing). */
+        Land, /**< @brief Landing. */
+        Offboard, /**< @brief In 'offboard' mode. */
+        FollowMe, /**< @brief In 'follow-me' mode. */
+        Manual, /**< @brief In 'Manual' mode. */
+        Altctl, /**< @brief In 'Altitude Control' mode. */
+        Posctl, /**< @brief In 'Position Control' mode. */
+        Acro, /**< @brief In 'Acro' mode. */
+        Stabilized, /**< @brief In 'Stabilize' mode. */
+    };
+
+    /**
+     * @brief Stream operator to print information about a `ActionServer::FlightMode`.
+     *
+     * @return A reference to the stream.
+     */
+    friend std::ostream& operator<<(std::ostream& str, ActionServer::FlightMode const& flight_mode);
+
+    /**
+     * @brief State to check if the vehicle can transition to
+     * respective flightmodes
+     */
+    struct AllowableFlightModes {
+        bool can_auto_mode{}; /**< @brief Auto/mission mode */
+        bool can_guided_mode{}; /**< @brief Guided mode */
```

Comment:
What do you mean with Guided mode? Guided is basically what we call Offboard in PX4, right?

Reply:
Yes I was basing this on the `base_mode` in mavlink first - it doesn't map to anything at the moment but offboard makes sense.